### PR TITLE
Add test suite

### PR DESCRIPTION
**Pull request by itsdalmo**
Based on the work in: https://github.com/buildkite/lifecycled/pull/50

I just decided to rewrite with inspiration from the PR above, to get a basic test suite up and running. In the process I figured it was prudent to do the following:

- Move `main.go` to `cmd/main.go` so that we can use package names to ensure that we are testing the public interface (i.e. whats usable from `main.go`) in our tests.
- Create a `logrus.Logger` to pass around, so that we can use `NullLogger` in the tests.
- Pass `log.Entry` to some listener and handler methods, so that we can propagate fields.
- Return the `TerminationNotice` from `daemon.Start()` so that we have a return value to test in our test suite. To not bloat `main.go` I decided to make a `Config` struct and have the `lifecycled.New()` and `NewDaemon()` functions do more of the boilerplate setup.

With the above in place, I wrote three basic tests to validate the behaviour of the daemon. So for future PR's etc we should be able to run tests to catch things like Queues not being cleaned up.

Possible improvements:
- Use the hook returned by the NullLogger to print the log entries when/if tests fail. Since we are handling a lot of errors by logging them, this would make it a lot easier to debug failing tests.

Let me know what you think 👍

### CODE REVIEW

**Comment by itsdalmo** [comment]
Thoughts on this @lox?

**Comment by lox** [comment]
Apologies, I've been sick! 😷Will review this tomorrow!

**Comment by itsdalmo** [comment]
NP! Get well soon 🙌

File: spot.go [hunk 1]
```diff
@@ -58,7 +60,7 @@ func (l *SpotListener) Start(ctx context.Context, notices chan<- TerminationNoti
 				log.Error("Empty response from metadata")
 				continue
 			}
-			t, err := time.Parse("2006-01-02T15:04:05Z", out)
+			t, err := time.Parse(time.RFC3339, out)
```
Comment (lox) [inline spot.go]: Are you sure this matches RFC3339? I recall pulling my hair out because it was a subtle variant.

**Comment by lox** [comment]
Can we go with `/cmd/lifecycled/main.go` so that `go install ./cmd/lifecycled` does the right thing?

File: spot.go (hunk 1 above)
Comment (itsdalmo) [reply spot.go]: I haven't tested this with an actual spot termination, but according to the docs it should work:
- RFC3339 is defined as `2006-01-02T15:04:05Z07:00` here: 
https://golang.org/pkg/time/#pkg-constants
- The EC2 metadata documentation also has an [example](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) of a spot termination timestamp:
 `2015-01-05T18:02:00Z`

The only gotcha that I'm aware of is that Go is strictly compliant with RFC3339, so it either accepts `Z` for Zulu time _or_ `+00:00`, but not both together (which I've seen sometimes, but not on AWS). That being said, after some quick tests the only difference `2006-01-02T15:04:05Z` and `time.RFC3339` is that `time.RFC3339` supports timestamps with offsets and not just Z for Zulu time.

tl;dr I think this should be fine, and we can always roll back if it does not work 👍

**Comment by itsdalmo** [comment]
@lox - Thanks for the review! Have moved the binary, and feel confident that `RFC3339` should work here - but we should test it in the wild before making a 3.0 release - are you up for that? 😄

**Comment by itsdalmo** [comment]
What do you think @lox - is this ready to be merged?
